Fix 51 issue

IlhamKhabibullin opened the pull request:
I think there is still a lot of work I have to do, now i want to get sure that i moving in right direction
@ffMathy you the one who I expect will assure me

IlhamKhabibullin commented:
@ffMathy I am planning to continue working on this PR after having your feedback

ffMathy commented:
Direction seems OK, if you can get the existing breaking tests passing.

Thank you so much for looking into this! <3

IlhamKhabibullin commented:
May be we should replace `received` method with some Symbol? That would free us from necessity of `disabledFor` stuff? 

I will try without it, but just think about that. 
Sequlize library uses this approach for operators in db queries, operators like $where.

IlhamKhabibullin commented:
@ffMathy I have made required fixes

"class method received" test was broken because I was throwing error in `set` of `receivedProxy`, now it does same thing as simple `proxy` of `context` 

Two tests for disableFor also was broken, it actually seems like they was not working correctly, get, set and apply of disableProxy was sending its `arguments` object into get, set and apply of `Context`, which ones was expect not same list of arguments, for example, `context.get` was expecting only `property`, but it was getting `target, this, property`, and сonsequently getting `target` where it was expecting `property`

I fixed that

also it seems like `received` called with `disableFor` was never recording as `getPropertyState` inside `InitialState`, not sure if it was broken always or I did that, but anyway I fixed that as well

waiting for your feedback 

[
![](https://images-na.ssl-images-amazon.com/images/I/71LFj9PUO0L._SL1500_.jpg)
](url)

ffMathy commented:
So all tests pass?

ffMathy commented:
If so, that's impressive!

ffMathy commented on src/states/SetPropertyState.ts:
```diff
@@ -54,6 +54,6 @@ export class SetPropertyState implements ContextState {
 
     get(context: Context, property: PropertyKey): undefined {
-        return void 0;
+        return void 0; // TODO:  probably better throw error since this should never happen
```
In that case, feel free to throw here

IlhamKhabibullin replied:
done

ffMathy commented:
Do all tests pass?

ffMathy commented:
Or is it just the tests for issue #51?

IlhamKhabibullin commented:
All tests pass, 39 exact number, 1 tests skipped, but not by me

ffMathy commented:
@IlhamKhabibullin the build fails on the build server.

> spec/issues/59.test.ts(16,31): error TS2554: Expected 0 arguments, but got 1.

IlhamKhabibullin commented:
researching on it

ср, 13 нояб. 2019 г. в 14:18, Mathias Lykkegaard Lorenzen <
notifications@github.com>:

> @IlhamKhabibullin <https://github.com/IlhamKhabibullin> the build fails
> on the build server.
>
> spec/issues/59.test.ts(16,31): error TS2554: Expected 0 arguments, but got
> 1.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ffMathy/FluffySpoon.JavaScript.Testing.Faking/pull/70?email_source=notifications&email_token=AEO27IMDDQIM3OO24EC3AV3QTPPCDA5CNFSM4JLVFTPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOED5ZNXY#issuecomment-553359071>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEO27INM5IITHSJJV7TIMYLQTPPCDANCNFSM4JLVFTPA>
> .
>

IlhamKhabibullin commented:
isn't it because of previous PR
https://github.com/ffMathy/FluffySpoon.JavaScript.Testing.Faking/commit/92f7e60b871714f60ace5bb3ddc7ce6d5e0908ac

ср, 13 нояб. 2019 г. в 14:22, Ильхам Хабибуллин <mailbox.ilham@gmail.com>:

> researching on it
>
> ср, 13 нояб. 2019 г. в 14:18, Mathias Lykkegaard Lorenzen <
> notifications@github.com>:
>
>> @IlhamKhabibullin <https://github.com/IlhamKhabibullin> the build fails
>> on the build server.
>>
>> spec/issues/59.test.ts(16,31): error TS2554: Expected 0 arguments, but
>> got 1.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/ffMathy/FluffySpoon.JavaScript.Testing.Faking/pull/70?email_source=notifications&email_token=AEO27IMDDQIM3OO24EC3AV3QTPPCDA5CNFSM4JLVFTPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOED5ZNXY#issuecomment-553359071>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEO27INM5IITHSJJV7TIMYLQTPPCDANCNFSM4JLVFTPA>
>> .
>>
>

IlhamKhabibullin commented:
author says that change adds an example of failing received with optional
arguments, but it looks more like imperfections of typescript

ср, 13 нояб. 2019 г. в 14:41, Ильхам Хабибуллин <mailbox.ilham@gmail.com>:

> isn't it because of previous PR
> https://github.com/ffMathy/FluffySpoon.JavaScript.Testing.Faking/commit/92f7e60b871714f60ace5bb3ddc7ce6d5e0908ac
>
> ср, 13 нояб. 2019 г. в 14:22, Ильхам Хабибуллин <mailbox.ilham@gmail.com>:
>
>> researching on it
>>
>> ср, 13 нояб. 2019 г. в 14:18, Mathias Lykkegaard Lorenzen <
>> notifications@github.com>:
>>
>>> @IlhamKhabibullin <https://github.com/IlhamKhabibullin> the build fails
>>> on the build server.
>>>
>>> spec/issues/59.test.ts(16,31): error TS2554: Expected 0 arguments, but
>>> got 1.
>>>
>>> —
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/ffMathy/FluffySpoon.JavaScript.Testing.Faking/pull/70?email_source=notifications&email_token=AEO27IMDDQIM3OO24EC3AV3QTPPCDA5CNFSM4JLVFTPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOED5ZNXY#issuecomment-553359071>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AEO27INM5IITHSJJV7TIMYLQTPPCDANCNFSM4JLVFTPA>
>>> .
>>>
>>

ffMathy commented:
But didn't you mention that the tests were passing? How did you get it running on your machine?

IlhamKhabibullin commented:
yeah, that code was added 8 days ago, and I did my fork earlier, so on my
fork's master I wasn't having this broken change

ср, 13 нояб. 2019 г. в 14:55, Mathias Lykkegaard Lorenzen <
notifications@github.com>:

> But didn't you mention that the tests were passing? How did you get it
> running on your machine?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ffMathy/FluffySpoon.JavaScript.Testing.Faking/pull/70?email_source=notifications&email_token=AEO27IPDLJIEFQE56NFV4STQTPTKHA5CNFSM4JLVFTPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOED54QXI#issuecomment-553371741>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEO27IOANXTRPSJYGUMPWX3QTPTKHANCNFSM4JLVFTPA>
> .
>

ffMathy commented:
Ah that makes sense then. Do you have the courage to try to fix the optional issue too? 🙇
